### PR TITLE
Add ShardRetentionPolicy to Prometheus API

### PR DESCRIPTION
**Pull request by mviswanathsai**
## Description

Picking up from #6198 (all credits to @ArthurSens). I just stepped in since he's a bit busy!

Partially addresses https://github.com/prometheus-operator/prometheus-operator/issues/4967, by implementing half of what was proposed [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/proposals/202310-shard-autoscaling.md#graceful-scale-down-of-prometheus-servers).

When Scaling down Prometheus servers, there are good chances of data loss because the Operator is not aware if the data stored in a Prometheus server is being backed up by a Long-Term-Storage solution. Users who want to avoid such situations might want to tell the operator to keep the instance alive for a while before safe deletion.

This PR introduces the new field ShardRetentionPolicy to the Prometheus CRD to address the problem mentioned above and in the linked proposal.

For the sake of keeping PRs small, in this PR I'm only worrying about pod retention when a retention period is set. Follow-up PRs will be opened with the goals:

Scale up will re-use already existing pod that was marked for future deletion
Scale up/down will correctly redistribute targets.
Retention duration is not taken into consideration right now. If ShardRetentionPolicy is set to Retain, we'll keep the shard alive forever.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note

```

### CODE REVIEW

**Comment by mviswanathsai** [comment]
CC: @simonpasquier @nicolastakashi @slashpai

File: pkg/prometheus/server/operator.go [hunk 1]
```diff
@@ -951,6 +962,21 @@ func (c *Operator) sync(ctx context.Context, key string) error {
...
 
+// As the ShardRetentionPolicy feature evolves, should delete will evolve accordingly.
+// For now, shouldDelete just returns the appropriate boolean based on the retention type.
+func (c *Operator) shouldDelete(p *monitoringv1.Prometheus) (bool, error) {
+	if !c.retentionPoliciesEnabled {
+		// Feature-gate is disabled, default behavior is always to delete.
+		return true, nil
+	}
+
+	if p.Spec.ShardRetentionPolicy.WhenScaled == monitoringv1.WhenScaledRetentionTypeDelete {
```
Comment (simonpasquier) [inline pkg/prometheus/server/operator.go:973]: we need to test for nil pointers.

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 2]
```diff
@@ -1115,6 +1127,23 @@ type PrometheusSpec struct {
...
+type WhenScaledRetentionType string
+
+var (
+	WhenScaledRetentionTypeRetain WhenScaledRetentionType = "Retain"
+	WhenScaledRetentionTypeDelete WhenScaledRetentionType = "Delete"
+)
+
+type ShardRetentionPolicy struct {
+	// WhenScaled defines the retention policy when the Prometheus shards are scaled down.
+	// When set to `Retain`, the operator will keep the pods from the scaled-down shard(s), so the data can still be queried.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1139]:
```suggestion
	// * `Delete`, the operator will delete the pods from the scaled-down shard(s).
	// * `Retain`, the operator will keep the pods from the scaled-down shard(s).
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 3]
```diff
@@ -1115,6 +1127,23 @@ type PrometheusSpec struct {
...
+	//
+	// Default: "Delete"
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1141]:
```suggestion
	// If not defined, the operator assumes the `` Delete` value.
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 4]
```diff
@@ -1115,6 +1127,23 @@ type PrometheusSpec struct {
...
+	// +kubebuilder:validation:Enum=Retain;Delete
+	// +required
+	WhenScaled WhenScaledRetentionType `json:"whenScaledDown"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1144]: we had a discussion on #6198 that we wanted to align with the statefulset spec which uses `whenScaled` already.
```suggestion
	WhenScaled *WhenScaledRetentionType `json:"whenScaled,omitempty"`
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 5]
```diff
@@ -1115,6 +1127,23 @@ type PrometheusSpec struct {
...
+	WhenScaledRetentionTypeDelete WhenScaledRetentionType = "Delete"
+)
+
+type ShardRetentionPolicy struct {
+	// WhenScaled defines the retention policy when the Prometheus shards are scaled down.
+	// When set to `Retain`, the operator will keep the pods from the scaled-down shard(s), so the data can still be queried.
+	//
+	// Default: "Delete"
+	// +kubebuilder:validation:Enum=Retain;Delete
+	// +required
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1143]: I would define the field as optional for now.
```suggestion
	// +optional
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 6]
```diff
@@ -1115,6 +1127,23 @@ type PrometheusSpec struct {
 	EnableAdminAPI bool `json:"enableAdminAPI,omitempty"`
 }
 
+type WhenScaledRetentionType string
+
+var (
+	WhenScaledRetentionTypeRetain WhenScaledRetentionType = "Retain"
+	WhenScaledRetentionTypeDelete WhenScaledRetentionType = "Delete"
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1134]: (nit)
```suggestion
	RetainWhenScaledRetentionType WhenScaledRetentionType = "Retain"
	DeleteWhenScaledRetentionType WhenScaledRetentionType = "Delete"
```

**Comment by mviswanathsai** [comment]
Not sure why the generate and format is failing.

**Comment by slashpai** [comment]
> Not sure why the generate and format is failing.

Because import updated

```
 import (
-	v1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
 )
 
 // ShardRetentionPolicyApplyConfiguration represents a declarative configuration of the ShardRetentionPolicy type for use
 // with apply.
 type ShardRetentionPolicyApplyConfiguration struct {
-	WhenScaled *v1.WhenScaledRetentionType `json:"whenScaled,omitempty"`
+	WhenScaled *monitoringv1.WhenScaledRetentionType `json:"whenScaled,omitempty"`
 }
 ```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 7]
```diff
@@ -996,6 +996,18 @@ type PrometheusSpec struct {
 	// Maximum number of bytes used by the Prometheus data.
 	RetentionSize ByteSize `json:"retentionSize,omitempty"`
 
+	// ShardRetentionPolicy defines the retention policy for the Prometheus shards.
+	// (Alpha) Using this field requires the 'PrometheusShardRetentionPolicy' feature gate to be enabled.
+	//
+	// The final goals for this feature can be seen at https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/proposals/202310-shard-autoscaling.md#graceful-scale-down-of-prometheus-servers,
+	// however, the feature is not yet fully implemented and has known limitations:
+	// * Scale up after scale down will not re-use the previously marked-for-deletion shards.
+	// * Target redistribution doesn't happen on scale up/down.
```
Comment (nicolastakashi) [inline pkg/apis/monitoring/v1/prometheus_types.go:1005]: Hmm I'm not sure, I think whenever we increase or decrease number of replicas the shards will be distributed, but the data won't.

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 8]
```diff
@@ -996,6 +996,18 @@ type PrometheusSpec struct {
...
+	// * Retention duration is not settable, for now shards are retained forever.
```
Comment (nicolastakashi) [inline pkg/apis/monitoring/v1/prometheus_types.go:1006]: I don't know if I follow this.
Are we retaining data forever? What does it means?

File: pkg/prometheus/server/operator.go [hunk 9]
```diff
@@ -951,6 +962,22 @@ func (c *Operator) sync(ctx context.Context, key string) error {
...
 
+// As the ShardRetentionPolicy feature evolves, should delete will evolve accordingly.
+// For now, shouldDelete just returns the appropriate boolean based on the retention type.
+func (c *Operator) shouldDelete(p *monitoringv1.Prometheus) (bool, error) {
+	if !c.retentionPoliciesEnabled {
+		// Feature-gate is disabled, default behavior is always to delete.
+		return true, nil
+	}
+
+	if p.Spec.ShardRetentionPolicy.WhenScaled != nil &&
```
Comment (nicolastakashi) [inline pkg/prometheus/server/operator.go:973]: I think this work as well.

```suggestion
	if ptr.Dref(p.Spec.ShardRetentionPolicy.WhenScaled, monitoringv1.DeleteWhenScaledRetentionType) == monitoringv1.DeleteWhenScaledRetentionType {
	return true, nil
}
```

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 8 above)
Comment (simonpasquier) [reply pkg/apis/monitoring/v1/prometheus_types.go:1006]: I read it as: "if you choose policy=retain, the shards are never removed"

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 10]
```diff
@@ -996,6 +996,18 @@ type PrometheusSpec struct {
 	// Maximum number of bytes used by the Prometheus data.
 	RetentionSize ByteSize `json:"retentionSize,omitempty"`
 
+	// ShardRetentionPolicy defines the retention policy for the Prometheus shards.
+	// (Alpha) Using this field requires the 'PrometheusShardRetentionPolicy' feature gate to be enabled.
+	//
+	// The final goals for this feature can be seen at https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/proposals/202310-shard-autoscaling.md#graceful-scale-down-of-prometheus-servers,
+	// however, the feature is not yet fully implemented and has known limitations:
+	// * Scale up after scale down will not re-use the previously marked-for-deletion shards.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1004]: I don't think that it's true either. IIUC the only thing not supported after this PR is merged is the "delete shards based on their retention period" capability.

File: pkg/prometheus/server/operator.go [hunk 11]
```diff
@@ -951,6 +962,22 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 	return nil
 }
 
+// As the ShardRetentionPolicy feature evolves, should delete will evolve accordingly.
+// For now, shouldDelete just returns the appropriate boolean based on the retention type.
+func (c *Operator) shouldDelete(p *monitoringv1.Prometheus) (bool, error) {
```
Comment (simonpasquier) [inline pkg/prometheus/server/operator.go:967]: If I read correctly, the function never returns an error.
```suggestion
func (c *Operator) shouldDelete(p *monitoringv1.Prometheus) bool {
```

File: pkg/prometheus/server/operator.go [hunk 12]
```diff
@@ -940,8 +942,17 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 			return
 		}
 
+		shouldDelete, err := c.shouldDelete(p)
```
Comment (simonpasquier) [inline pkg/prometheus/server/operator.go:945]: (nit) it might be easier to read if the function is `shouldRetain` (e.g. no need to negate the returned value to exit early).

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 7 above)
Comment (mviswanathsai) [reply pkg/apis/monitoring/v1/prometheus_types.go:1005]: I think Arthur just wanted to list out what is/isn't possible after this PR alone. I may be wrong though.

>I think whenever we increase or decrease number of replicas the shards will be distributed, but the data won't.
Yep, Agreed.

Comment (mviswanathsai) [reply pkg/apis/monitoring/v1/prometheus_types.go:1005]: I will update it a bit to reflect the current changes.

File: pkg/prometheus/server/operator.go (hunk 11 above)
Comment (mviswanathsai) [reply pkg/prometheus/server/operator.go:967]: I think it was added to "future proof" this function. Later on, `shouldDelete` would also involve asking the API server for the object and getting the deletion annotation.
We can leave this too, or remove it if it seems to cause confusion.

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 10 above)
Comment (mviswanathsai) [reply pkg/apis/monitoring/v1/prometheus_types.go:1004]: >IIUC the only thing not supported after this PR is merged is the "delete shards based on their retention period" capability.

Correct

File: pkg/prometheus/server/operator.go (hunk 11 above)
Comment (simonpasquier) [reply pkg/prometheus/server/operator.go]: we can always change the function signature later. I'm more on the side of keeping it consistent (e.g. if it can't return an error, it shouldn't have an error return value).

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 13]
```diff
@@ -1128,6 +1138,24 @@ type PrometheusSpec struct {
...
 
+type WhenScaledRetentionType string
+
+var (
+	RetainWhenScaledRetentionType WhenScaledRetentionType = "Retain"
+	DeleteWhenScaledRetentionType WhenScaledRetentionType = "Delete"
+)
+
+type ShardRetentionPolicy struct {
+	// WhenScaled defines the retention policy when the Prometheus shards are scaled down.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1149]: (nit)
```suggestion
	// Defines the retention policy when the Prometheus shards are scaled down.
```

File: pkg/prometheus/server/operator.go [hunk 14]
```diff
@@ -940,8 +942,17 @@ func (c *Operator) sync(ctx context.Context, key string) error {
...
+			c.logger.Error("failed to determine if StatefulSet should be retained", "err", err, "name", s.GetName(), "namespace", s.GetNamespace())
+			return
+		}
+		if shouldRetain {
+			return
+		}
+
 		if err := ssetClient.Delete(ctx, s.GetName(), metav1.DeleteOptions{PropagationPolicy: ptr.To(metav1.DeletePropagationForeground)}); err != nil {
-			c.logger.Error("failed to delete StatefulSet object", "err", err, "name", s.GetName(), "namespace", s.GetNamespace())
+			c.logger.Error("deleting StatefulSet failed", "err", err, "name", s.GetName(), "namespace", s.GetNamespace())
```
Comment (simonpasquier) [inline pkg/prometheus/server/operator.go:955]: (nit) I would retain the current form.

**Review by simonpasquier — APPROVED** [review]
thanks!
